stacks-signer-api/src/db/keys.rs

```diff
@@ -0,0 +1,308 @@
+use crate::{
+    db::Error,
+    key::Key,
+    routes::keys::KeysQuery,
+    signer::{Signer, Status},
+};
+
+use sqlx::{Row, SqlitePool};
+use warp::http;
+
+use super::signers::add_signer;
+
+const fn sql_insert_key() -> &'static str {
+    r#"
+    INSERT OR REPLACE INTO keys (signer_id, user_id, key) VALUES (?1, ?2, ?3)
+    "#
+}
+
+const fn sql_delete_key() -> &'static str {
+    r#"
+    DELETE FROM keys WHERE signer_id = ?1 AND user_id = ?2 AND key = ?3
+    "#
+}
+
+const fn sql_delete_keys_by_id() -> &'static str {
+    r#"
+    DELETE FROM keys WHERE signer_id = ?1 AND user_id = ?2
+    "#
+}
+
+const fn sql_select_keys() -> &'static str {
+    r#"
+    SELECT key FROM keys WHERE signer_id = ?1 AND user_id = ?2 ORDER BY key ASC
+    "#
+}
+
+const fn sql_count_keys_by_id() -> &'static str {
+    r#"
+    SELECT COUNT(*) FROM keys WHERE signer_id = ?1 AND user_id = ?2
+    "#
+}
+
+pub async fn add_key(key: Key, pool: SqlitePool) -> Result<impl warp::Reply, warp::Rejection> {
+    // First make sure we have an existing signer id;
+    let count: i64 = sqlx::query(sql_count_keys_by_id())
+        .bind(key.signer_id)
+        .bind(key.user_id)
+        .fetch_one(&pool)
+        .await
+        .map_err(Error::from)?
+        .get(0);
+
+    if count == 0 {
+        let signer = Signer {
+            signer_id: key.signer_id,
+            user_id: key.user_id,
+            status: Status::Inactive,
+        };
+        add_signer(signer, pool.clone()).await?;
+    }
+
+    sqlx::query(sql_insert_key())
+        .bind(key.signer_id)
+        .bind(key.user_id)
+        .bind(key.key.as_str())
+        .execute(&pool)
+        .await
+        .map_err(Error::from)?;
+
+    Ok(warp::reply::with_status(
+        warp::reply::json(&serde_json::json!({ "status": "added" })),
+        http::StatusCode::CREATED,
+    ))
+}
+
+pub async fn delete_keys_by_id(
+    signer_id: i64,
+    user_id: i64,
+    pool: SqlitePool,
```

Comment:
Although an Arc is cheap to clone a reference is still cheaper. There's no reason for this function to own the pool variable. We can save some reference count mutations and a few `.clone()` calls by taking a reference here (and in the other functions) instead.

```suggestion
    pool: &SqlitePool,
```

Reply:
These all take a SqlitePool because I could not  satisfy the warp api/lifetimes without passing pool.clone() which then forces the subsequent function signature to take the object rather than a reference. I can try again, but I have my doubts that it is possible.